### PR TITLE
Maths tests resurrected (exp and log functions) with bugfixes

### PR DESCRIPTION
**Pull request by LewisGaul**
New PR following bad merge on https://github.com/ziglang/zig/pull/16225 (previously reviving https://github.com/ziglang/zig/pull/13311, https://github.com/ziglang/zig/pull/12355 and https://github.com/ziglang/zig/pull/10415!).

- Add test framework (at `lib/std/math/test.zig`) for maths functions that take in a float and output a float of the same type, e.g. `exp()`
- Create tests for all exp/log functions under `lib/std/math/test/` and `lib/compiler_rt/test/` (alongside implementations under test)
  - Sanity tests, taken from [libc-test](https://wiki.musl-libc.org/libc-test.html) (e.g. [exp.h](http://nsz.repo.hu/git/?p=libc-test;a=blob;f=src/math/sanity/exp.h;h=7cbdfcae8770955be175e824e92e535fabae4430;hb=HEAD))
  - Special cases (inf, nan, 0, ...)
  - Boundary cases (last value before inf, first subnormal, ...)
- Remove special case tests from function implementation modules to avoid duplication (leave the basic tests intact)
- Bugfixes uncovered by improved testing:
  - `exp2_32()` flushing to zero too soon
  - `expm1_32()` giving incorrect results for a certain range of small negative inputs
  - `ln64()` giving incorrect results for positive subnormal inputs
  - Some functions returning canonical NaN instead of passing through input NaNs unchanged (as per Musl and GCC)
  - `log10_64()` returning 32-bit inf and NaN, which would be fine as it's cast up, except casting NaN results in failure to pass through the input unchanged

Need resolving:
- [ ] Allow compiler-rt to access the maths test utils without exposing it from the std package available to all users (see https://github.com/ziglang/zig/pull/16808/files#r1292880458 and the following paragraphs)
- [X] Merge https://github.com/ziglang/zig/pull/16826 and rebase, to remove use of signalling NaNs

Some history/context:
- When I originally wrote these tests all the code under test was in `lib/std/math/`, so it made sense for the test infra to live there too (it was suggested to me to use `lib/std/math/test.zig` and `lib/std/math/test/`).
- Some of the code under test now lives in `lib/compiler_rt/`, with the functions in `lib/std/math/` only falling back to the `compiler_rt` code *sometimes* (depending on the target, I'm not clear exactly when).
- We still want the tests to live alongside the code under test, i.e. some tests in `lib/std/math/` and others in `lib/compiler_rt/`.
- The test infra in `lib/std/math/test.zig` is still relevant across all of these tests.

The one remaining problem is I don't know how to share this test infra between `lib/compiler_rt/` and `lib/std/math/`... Currently I have exposed `lib/std/math/test.zig` as `std.math.test_utils`, but I don't think this is desirable! If I try to import the file directly from tests in `lib/compiler_rt/test/` with `@import("../../std/math/test.zig");` then I get the compiler error "error: file exists in multiple modules" :( If it can't be shared between `std.math` tests and `compiler_rt` tests without exposing it to users, perhaps it should be moved to somewhere like `lib/testing/`?

Example output with 'log_level = .info' for debugging (failing test uncommented for illustration):
```
$./zig test lib/compiler_rt.zig --test-filter "math.exp64() special"
Test [31/31] test.math.exp64() special... [default] (info):
[default] (info):  IN:  0x0000000000000000  0x0.0p0                   0.0e+00
[default] (info): OUT:  0x3FF0000000000000  0x1p0                     1.0e+00
[default] (info):
[default] (info):  IN:  0x0000000000000000  0x0.0p0                   0.0e+00
[default] (info): OUT:  0x3FF0000000000000  0x1p0                     1.0e+00
[default] (info):
[default] (info):  IN:  0x3FF0000000000000  0x1p0                     1.0e+00
[default] (info): OUT:  0x4005BF0A8B14576A  0x1.5bf0a8b14576ap1       2.7182818284590455e+00
[default] (info): EXP:  0x4005BF0A8B145769  0x1.5bf0a8b145769p1       2.718281828459045e+00
FAILURE: expected exp(0x1p0)->0x1.5bf0a8b145769p1, got 0x1.5bf0a8b14576ap1 (64-bit)
         expected exp(0x3FF0000000000000)->0x4005BF0A8B145769, got 0x4005BF0A8B14576A
[default] (info):
[default] (info):  IN:  0x3FE62E42FEFA39EF  0x1.62e42fefa39efp-1      6.931471805599453e-01
[default] (info): OUT:  0x4000000000000000  0x1p1                     2.0e+00
[default] (info):
[default] (info):  IN:  0x7FF0000000000000  inf                       inf
[default] (info): OUT:  0x7FF0000000000000  inf                       inf
[default] (info):
[default] (info):  IN:  0xFFF0000000000000  -inf                      -inf
[default] (info): OUT:  0x0000000000000000  0x0.0p0                   0.0e+00
[default] (info):
[default] (info):  IN:  0x7FF0000000000001  nan                       nan
[default] (info): OUT:  0x7FF0000000000001  nan                       nan
[default] (info):
[default] (info):  IN:  0xFFF0000000000001  -nan                      -nan
[default] (info): OUT:  0xFFF0000000000001  -nan                      -nan
[default] (info):
[default] (info):  IN:  0x7FF0123400000000  nan                       nan
[default] (info): OUT:  0x7FF0123400000000  nan                       nan
[default] (info):
[default] (info):  IN:  0xFFF0123400000000  -nan                      -nan
[default] (info): OUT:  0xFFF0123400000000  -nan                      -nan
[default] (info):
[default] (info): Subtest summary: 9 passed; 1 failed
Test [31/31] test.math.exp64() special... FAIL (Failure)
/mnt/c/Users/legaul/repos/zig/lib/std/math/test.zig:114:23: 0x2448bc in runTests__anon_8139 (test)
    if (failures > 0) return error.Failure;
                      ^
/mnt/c/Users/legaul/repos/zig/lib/compiler_rt/test/exp.zig:116:5: 0x24492b in test.math.exp64() special (test)
    try runTests(cases);
    ^
30 passed; 0 skipped; 1 failed.
error: the following test command failed with exit code 1:
/mnt/c/Users/legaul/repos/zig/zig-cache/o/53dd19fa852e86f59f08ef2154127c98/test
```

---

Previous suggestions from @matu3ba that I'm considering out of scope of this PR:

- Adding tests for all extreme values (e.g. +/- inf) rather than targeted boundary cases (see https://github.com/ziglang/zig/pull/10415#discussion_r775308449)

- Taking on all Musl tests
> 2 ideas for discussion
> 
> 1. port [libc-test](https://wiki.musl-libc.org/libc-test.html) to zig
> 
> * 1.1. Port the test running functions from [libc-test](https://wiki.musl-libc.org/libc-test.html), which is the test library of musl.
> * 1.2. Use scripts to generate the test code from the musl test macros.
> * more code to write, but less dependency on C/clang
> * caveat, but actually a feature: Include the comments in the generated test code.
> 
> 2. vendor [libc-test](https://wiki.musl-libc.org/libc-test.html)
> 
> * 2.1 use zig stubs for the build scripts and test calls
> * less code to write, but more dependency on C/clang

### CODE REVIEW

File: lib/std/math.zig [hunk 1]
```diff
@@ -257,6 +259,10 @@ pub const cosh = @import("math/cosh.zig").cosh;
 pub const tanh = @import("math/tanh.zig").tanh;
 pub const gcd = @import("math/gcd.zig").gcd;
 
+// TODO: This should not be part of the public API, but compiler-rt tests need
+//       access to it...
+pub const test_utils = @import("math/test.zig");
```
Comment (LewisGaul) [inline lib/std/math.zig:264]: This needs resolving and removing

**Comment by matu3ba** [comment]
almost, only NaN stuff on linux x86 debug and release are failing with for example:
```
FAILURE: expected exp(-nan)->-nan, got -nan (64-bit)
FAILURE: expected exp(nan)->nan, got nan (64-bit)
FAILURE: expected exp(-nan)->-nan, got -nan (64-bit)
```

I think you are running here into x86 float NaN copies being broken https://bugs.llvm.org/show_bug.cgi?id=45152 linked from https://discourse.llvm.org/t/semantics-of-nan/66729 by Ralf:

"It would also be interesting to know whether just copying floats around guarantees to preserve NaN bits or not. I.e., can only float arithmetic change NaNs, or can this even happen “silently”? I sure hope only operations can affect NaN values but at least on some architectures, that [does not currently seem to be guaranteed on 32bit x86](https://bugs.llvm.org/show_bug.cgi?id=45152). Would be good to get an official statement whether this is a bug/deficiency of that particular backend, or part of the LLVM semantics."

Can you double check, if the x86 float is copied around or if the call is inlined?

File: lib/std/math/nan.zig [hunk 2]
```diff
@@ -2,6 +2,18 @@ const math = @import("../math.zig");
...
 pub inline fn nan(comptime T: type) T {
+    return switch (@typeInfo(T).Float.bits) {
+        16 => math.qnan_f16,
+        32 => math.qnan_f32,
+        64 => math.qnan_f64,
+        80 => math.qnan_f80,
+        128 => math.qnan_f128,
+        else => @compileError("unreachable"),
+    };
+}
```
Comment (LewisGaul) [inline lib/std/math/nan.zig:13]: This change should be incorporated via the separate PR https://github.com/ziglang/zig/pull/16826.

**Comment by LewisGaul** [comment]
> The one remaining problem is I don't know how to share this test infra between `lib/compiler_rt/` and `lib/std/math/`... Currently I have exposed `lib/std/math/test.zig` as `std.math.test_utils`, but I don't think this is desirable! If I try to import the file directly from tests in `lib/compiler_rt/test/` with `@import("../../std/math/test.zig");` then I get the compiler error "error: file exists in multiple modules" :( If it can't be shared between `std.math` tests and `compiler_rt` tests without exposing it to users, perhaps it should be moved to somewhere like `lib/testing/`?

Could anyone help with a recommendation here? When I previously suggested moving this test infra to an internal `lib/testing/` dir @andrewrk thought this was overkill, but I can't see another way to do it without making this part of what's publicly available from `std`, which seems undesirable to me.

**Comment by LewisGaul** [comment]
> The one remaining problem is I don't know how to share this test infra between `lib/compiler_rt/` and `lib/std/math/`... Currently I have exposed `lib/std/math/test.zig` as `std.math.test_utils`, but I don't think this is desirable! If I try to import the file directly from tests in `lib/compiler_rt/test/` with `@import("../../std/math/test.zig");` then I get the compiler error "error: file exists in multiple modules" :( If it can't be shared between `std.math` tests and `compiler_rt` tests without exposing it to users, perhaps it should be moved to somewhere like `lib/testing/`?

Could anyone help with a recommendation here? When I previously suggested moving this test infra to an internal `lib/testing/` dir @andrewrk thought this was overkill, but I can't see another way to do it without making this part of what's publicly available from `std`, which seems undesirable to me.

**Comment by andrewrk** [comment]
Ah, finally, you left one on a successful CI run. Let me see if I can help you get this landed, and help you resolve the testing problem you mentioned.

I've been struggling to understand why there is a problem at all because it seems like you either need to test it in compiler_rt, or in std, but not both. I'll take a deeper look and see if I can help.

**Comment by andrewrk** [comment]
After updating to the latest zig, I'm getting std lib test failures:

```
Test [793/2640] test.math.expm1_32() special... FAILURE: expected expm1(-0x0.0p0)->0x0.0p0, got -0x0.0p0 (32-bit)
         expected expm1(0x80000000)->0x00000000, got 0x80000000
Test [793/2640] test.math.expm1_32() special... FAIL (Failure)
/home/andy/Downloads/zig/lib/std/math/test.zig:119:23: 0x1a517c2 in runTests__anon_39188 (test)
    if (failures > 0) return error.Failure;
                      ^
/home/andy/Downloads/zig/lib/std/math/test/expm1.zig:51:5: 0x1a5181b in test.math.expm1_32() special (test)
    try runTests(cases);
    ^
Test [796/2640] test.math.expm1_64() special... FAILURE: expected expm1(-0x0.0p0)->0x0.0p0, got -0x0.0p0 (64-bit)
         expected expm1(0x8000000000000000)->0x0000000000000000, got 0x8000000000000000
Test [796/2640] test.math.expm1_64() special... FAIL (Failure)
/home/andy/Downloads/zig/lib/std/math/test.zig:119:23: 0x1a522de in runTests__anon_39240 (test)
    if (failures > 0) return error.Failure;
                      ^
/home/andy/Downloads/zig/lib/std/math/test/expm1.zig:94:5: 0x1a5232b in test.math.expm1_64() special (test)
    try runTests(cases);
    ^
```

File: lib/compiler_rt/test/exp.zig [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        tc32(-0x1.ab6d70p+3, 0x1.a88adep-20),
+        tc32(-0x1.5ac18ep+2, 0x1.22b328p-8 ),
+        tc32(-0x1.925982p-1, 0x1.d2acc0p-2 ),
+        tc32( 0x1.7221cep+3, 0x1.9c2ceap+16),
+        tc32( 0x1.11a0d4p+4, 0x1.980ee6p+24),
+        tc32(-0x1.ae41a2p+1, 0x1.1c28d0p-5 ),
+        tc32(-0x1.329154p+4, 0x1.47ef94p-28),
+        // zig fmt: on
+    };
+    try runTests(cases);
```
[the remark this answers is not in the document]
Comment (LewisGaul) [reply lib/compiler_rt/test/exp.zig:59]: In general this is a style preference (where the Zig codebase preference is now clear to me), but in this case it was an attempt to achieve concrete goals not natively available with standard Zig tests (which I can only assume are either being misunderstood or simply considered unworthy goals). I'll elaborate in another reply, but I appreciate this is a highly unimportant detail in the scope of the project so if I get a chance I'll rewrite as suggested and have this test setup live in my own repo for when I might need the testing features I implemented.

File: lib/compiler_rt/log.zig [hunk 4]
```diff
@@ -38,6 +38,11 @@ pub fn logf(x_: f32) callconv(.C) f32 {
     const Lg3: f32 = 0x91e9ee.0p-25;
     const Lg4: f32 = 0xf89e26.0p-26;
 
+    // TODO: This should be handled beneath.
```
[the remark this answers is not in the document]
Comment (LewisGaul) [reply lib/compiler_rt/log.zig:41]: I was considering this out of scope of this PR - I was just leaving the NaN functionality working and consistent without tidying up the existing TODOs like this (which I was prepared to look at afterwards back when I originally worked on this).

**Comment by LewisGaul** [comment]
> It's so close. Don't overthink it. Just get the bugs fixed and make sure they have coverage. If you try to do too much it's going to delay the PR getting merged again.

I can fully understand that perspective, although I don't think the contents of this PR are a symptom of overthinking, perhaps more a lack of familiarity with approach used in the Zig project (which I don't take issue with, it's just an observation).

I was seeing the value-add here as the testsuite, more than fixing the bugs. I wrote the testsuite to allow me to provide testing for the f128 support I implemented for these functions (which I completed, but was waiting on getting this PR in first!).

My goals for the tests were:
1. Declarative input and expected output (mirroring the musl approach)
2. Testing all declared inputs even when there are failures, with clear failure messages (including exact bit patterns)
3. Test multiple inputs per testcase (otherwise the test code would be very verbose and wouldn't really achieve point 1)
4. Shared definition for cases that are common across float size and/or function under test, to avoid mistakes (e.g. how NaN values should behave)

I don't think any of these are really possible if I simplify in the way you've requested :(

Anyway, I'll see if I can quickly transform the tests into the simple feature-free approach - something is better than nothing and I don't want to waste any more of your time in discussing/reviewing this low-priority work. Just wanted to provide my reasoning for going this way in the first place!
